packages/core/src/indexing/query-filter-parser.ts

```diff
@@ -0,0 +1,159 @@
+import {
+  QueryFilters as ApiQueryFilters,
+  ObjectFilter as ApiObjectFilter,
+} from '@ceramicnetwork/common'
+
+export type ValueFilterType = 'boolean' | 'number' | 'string'
+export type ValueFilter<T extends boolean | number | string = boolean | number | string> =
+  | { type: ValueFilterType; op: 'null'; value: boolean }
+  | { type: ValueFilterType; op: '='; value: T }
+  | { type: ValueFilterType; op: '!='; value: T }
+
+export type NonBooleanValueFilterType = 'number' | 'string'
+export type NonBooleanValueFilter<T extends number | string = number | string> =
+  | { type: NonBooleanValueFilterType; op: 'in'; value: Array<T> }
+  | { type: NonBooleanValueFilterType; op: 'nin'; value: Array<T> }
+  | { type: NonBooleanValueFilterType; op: '>'; value: T }
+  | { type: NonBooleanValueFilterType; op: '>='; value: T }
+  | { type: NonBooleanValueFilterType; op: '<'; value: T }
+  | { type: NonBooleanValueFilterType; op: '<='; value: T }
+
+export type AllValueFilter = ValueFilter | NonBooleanValueFilter
+
+/**
+ * Mapping of object keys to value filter
+ */
+export type ObjectFilter = Record<string, AllValueFilter>
+
+/**
+ * Advanced query filters on a document fields
+ */
+export type QueryFilters =
+  | { type: 'where'; value: ObjectFilter }
+  | { type: 'and'; value: Array<QueryFilters> }
+  | { type: 'or'; value: Array<QueryFilters> }
+  | { type: 'not'; value: QueryFilters }
+
+function getNonBooleanValueType<T extends number | string>(value: T): NonBooleanValueFilterType {
+  if (typeof value == 'number') {
+    return 'number'
+  } else {
+    return 'string'
+  }
+}
+
+function getValueType<T extends boolean | number | string>(value: T): ValueFilterType {
+  if (typeof value == 'boolean') {
+    return 'boolean'
+  } else if (typeof value == 'number') {
+    return 'number'
+  } else {
+    return 'string'
+  }
+}
+export function parseObjectFilter(filter: ApiObjectFilter): ObjectFilter {
+  const entries: Record<string, AllValueFilter> = {}
+  const keys = Object.keys(filter)
+  for (const key of keys) {
+    const childFilter = filter[key]
+    if ('isNull' in childFilter) {
```

Comment:
this doesn't actually do what we discussed on the call of using type assertions to do these checks so we can leverage the type system more natively.  That said I don't personally feel strongly about this so I'm fine with it if it works for @PaulLeCam and @bigs